### PR TITLE
Entropy estimation with non-invertible generative models

### PR DESCRIPTION
**Pull request by austin-hoover**
This PR adds the class `NNDist` to the `gpsr.beams` module. With a non-invertible transformation from the base distribution (`NNTransform`), we can't evaluate the probability density $p(x)$ at an arbitrary point $x$. But we can still evaluate the density at a batch of sampled points $\\{ x^{(i)} \\} \sim p(x)$ by computing the Jacobian matrix of the transformation. This can provide differentiable estimates of the KL divergence or other integrals involving the distribution function.

## Changes

* Added method `compute_batched_jacobian(x: torch.Tensor, function: Callable) -> torch.Tensor:` to `gpsr.beams` module. This method computes the Jacobian matrix of callable `function` for a batch of particles `x`, following the example [here](https://docs.pytorch.org/functorch/stable/generated/functorch.jacrev.html).
* Added class `gpsr.beams.NNDist`. This is a subclass of `GenModel` that uses the `NNTransform` as the transformation from the base distribution. It has methods `sample` and `sample_and_log_prob` but not `log_prob`.
* Renamed class `gpsr.beams.NSF` to `gpsr.beams.NSFDist` for consistency. Also changed arguments in `NSFDist` to match `NNDist` (`width`, `depth`).
* Added method `entropy(n: int, prior: Any) -> torch.Tensor` to the `gpsr.beams.GenModel` class. This is a convenience method to estimate the entropy or KL divergence from a batch of `n` samples, using the `sample_and_log_prob` method.
* Added [examples](https://github.com/austin-hoover/gpsr/tree/jacobian/docs/examples/kl) which train the model by minimizing the KL divergence.

test

### CODE REVIEW

File: docs/examples/jacobian/test_jacobian.ipynb
Comment (roussel-ryan) [inline docs/examples/jacobian/test_jacobian.ipynb:1]: Can we move this to a unit test instead of having a python notebook?

File: docs/examples/jacobian/test_kl_nn.ipynb
Comment (roussel-ryan) [inline docs/examples/jacobian/test_kl_nn.ipynb:1]: this is closer to a usable example, so we should rename to kl_nn.ipynb. A separate unit test would be useful for validating changes to beams.py

**Review by roussel-ryan — COMMENTED** [review]
Really cool stuff @austin-hoover . Is there a reference for the batched jacobian method of estimating the entropy using this technique?

**Comment by austin-hoover** [comment]
For references: the formula for KL divergence is the same as for normalizing flows (Section 2.3.2 of https://arxiv.org/pdf/2404.00502). There's some discussion about computing Jacobian determinant for non-invertible models [here](https://arxiv.org/pdf/2404.00502).

File: docs/examples/jacobian/kl_nn.ipynb
Comment (austin-hoover) [reply docs/examples/jacobian/kl_nn.ipynb:1]: Renamed kl_nn.ipynb

File: docs/examples/jacobian/test_jacobian.ipynb
Comment (austin-hoover) [reply docs/examples/jacobian/test_jacobian.ipynb:1]: I added methods to `test_beams.py` here: https://github.com/austin-hoover/gpsr/blob/913c247477011841ba6d1bf3a47c162b9956a63a/gpsr/tests/test_beams.py#L40-L71. The first one tests that Jacobian of `torch.sin` is correct. The second one computes the Jacobian of the `NNTransform` forward method.

**Comment by austin-hoover** [comment]
Marking this as a draft for now.

* The KL minimization looks like it's working in 2D examples, but convergence is slower than the normalizing flow and sometimes fails to match the target distribution, depending on initialization. So far I can't reproduce correct MaxEnt solutions for 2D reconstruction.
* I also tried to train the NN using a sliced Wasserstein loss function to regress to a target distribution. In this case, the NN has an even harder time converging. So maybe the log-prob calculation is fine and these issues have to do with the specific NN architecture used.

**Comment by austin-hoover** [comment]
Added test under /gpsr/tests/test_beams.py that verifies the log_prob calculation is the same as the exact result from the normalizing flow.

**Comment by roussel-ryan** [comment]
Thanks @austin-hoover. I'm looking into the testing errors now since its effecting the main branch
